File: data/stream_ohlc.py
from datetime import datetime

from pyspark import SparkConf, SparkContext
from pyspark.streaming import StreamingContext
from pyspark.sql import functions as F
from pyspark.sql import Row, SQLContext
from pyspark.sql.types import TimestampType


conf = SparkConf() \
        .setAppName('OHLC') \
        .setMaster('local[4]') \
        .set('spark.cassandra.connection.host', '10.0.2.15')

sc = SparkContext(conf=conf)
ssc = StreamingContext(sc, 30)

transactions = ssc.socketTextStream('localhost', 9999)

def get_sql_context_instance(sparkContext):
    if ('sparkSessionSingletonInstance' not in globals()):
        globals()['sparkSessionSingletonInstance'] = SQLContext(sparkContext)
    return globals()['sparkSessionSingletonInstance']

truncate_min = F.udf(lambda dt: datetime(dt.year, dt.month, dt.day, dt.hour, dt.minute), TimestampType())

def reduce_to_ohlc(time, rdd):
    row_rdd = rdd.map(lambda row: row.split(',')) \
                 .filter(lambda row: len(row) == 3) \
                 .map(lambda row: Row(
                       symbol=row[0],
                       tx_time=datetime.strptime(row[2], '%Y-%m-%d %H:%M:%S.%f'),
                       price=float(row[1])
                 ))
    sql_context = get_sql_context_instance(rdd.context)
    data = sql_context.createDataFrame(row_rdd)
    data.cache()
    data.write.format('org.apache.spark.sql.cassandra') \
            .options(table='transactions2', keyspace='stock', cluster='Test Cluster') \
            .mode('append') \
            .save()

    ohlc = data.select('symbol', truncate_min(data.tx_time).alias('batch_time'), 'price', 'tx_time') \
                .orderBy('tx_time') \
                .groupBy('symbol', 'batch_time') \
                .agg(
                   F.first(data.price).alias('open'),
                   F.max(data.price).alias('high'),
                   F.min(data.price).alias('low'),
                   F.last(data.price).alias('close'),
                   F.first(data.tx_time).alias('open_time'),
                   F.last(data.tx_time).alias('close_time')
                )

    existing_ohlc = sql_context.read.format('org.apache.spark.sql.cassandra') \
            .options(table='ohlc_1_min2', keyspace='stock', cluster='Test Cluster') \
            .load() \
            .select('symbol', 'batch_time', 'open', 'open_time', 'high', 'low', 'close', 'close_time')

    merged_ohlc = ohlc.join(existing_ohlc,
                             (ohlc.symbol == existing_ohlc.symbol) &
                             (ohlc.batch_time == existing_ohlc.batch_time),
                             'left'
                           )

    merged_ohlc = merged_ohlc.select(
        ohlc.symbol.alias('symbol'),
        ohlc.batch_time.alias('batch_time'),
        F.when(existing_ohlc.open_time < ohlc.open_time, existing_ohlc.open).otherwise(ohlc.open).alias('open'),
        F.when(existing_ohlc.open_time < ohlc.open_time, existing_ohlc.open_time).otherwise(ohlc.open_time).alias('open_time'),
        F.when(existing_ohlc.close_time > ohlc.close_time, existing_ohlc.close).otherwise(ohlc.close).alias('close'),
        F.when(existing_ohlc.close_time > ohlc.close_time, existing_ohlc.close_time).otherwise(ohlc.close_time).alias('close_time'),
        F.when(existing_ohlc.low < ohlc.low, existing_ohlc.low).otherwise(ohlc.low).alias('low'),
        F.when(existing_ohlc.high > ohlc.high, existing_ohlc.high).otherwise(ohlc.high).alias('high')
    )
    merged_ohlc.write.format('org.apache.spark.sql.cassandra') \
                .options(table='ohlc_1_min2', keyspace='stock', cluster='Test Cluster') \
                .mode('append') \
                .save()


transactions.foreachRDD(reduce_to_ohlc)
ssc.start()
ssc.awaitTermination()

File: data/stream_ohlc/src/main/scala/stream_ohlc.scala
import java.text.SimpleDateFormat
import java.util.{Date, Calendar}

import org.apache.spark._
import org.apache.spark.streaming._
import com.datastax.spark.connector._
import com.datastax.spark.connector.streaming._

object StreamingOhlc {
  val conf = new SparkConf()
              .setMaster("local[2]")
              .setAppName("OHLC")

  val ssc = new StreamingContext(conf, Seconds(20))


  def refreshOhlc(key: (String, Calendar), value: Option[OhlcValue], state: State[OhlcValue]) = {
    var newOhlcValue: OhlcValue = null
    if (state.exists) {
      val existingState: OhlcValue = state.get
      if (value.isDefined) {
        newOhlcValue = reduceOhlc(value.get, existingState)
        state.update(newOhlcValue)
      } else {
        newOhlcValue = existingState
        val expiredTime = Calendar.getInstance
        expiredTime.roll(Calendar.MINUTE, -3)
        if (key._2.before(expiredTime))
          state.remove
      }
    } else {
      newOhlcValue = value.get
      state.update(newOhlcValue)
    }

    (key, newOhlcValue)
  }

  val refreshOhlcSpec = StateSpec.function(refreshOhlc _)


  def main(args: Array[String]) {
    ssc.checkpoint("_checkpoints")
    
    //TODO: Thread-safe aware
    val formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS")
    def parseDateTime(t: String) = {
      val cal = Calendar.getInstance()
      cal.setTime(formatter.parse(t))
      cal
    }

    val transactions = ssc.socketTextStream("localhost", 9999)
                    .map(_.split(','))
                    .map(t => (t(0), parseDateTime(t(2)), t(1).toFloat))

    transactions.cache
    transactions.saveToCassandra("stock", "transactions2",
      SomeColumns("symbol", "tx_time", "price"))

    val ohlcStream = transactions.transform(rdd => {
      def truncateMin(cal: Calendar): Calendar = {
        val newCal: Calendar = cal.clone().asInstanceOf[Calendar]
        newCal.set(Calendar.SECOND, 0)
        newCal.set(Calendar.MILLISECOND, 0)
        newCal
      }

      rdd.map(t => new BatchTransaction(t._1, truncateMin(t._2), t._3, t._2))
        .keyBy(t => (t.symbol, t.batchTime))
        .mapValues(t => new OhlcValue(t.price, t.price, t.price, t.price, t.txTime, t.txTime))
        .reduceByKey(reduceOhlc _)
    }).mapWithState(refreshOhlcSpec)

    ohlcStream.map(x => (x._1._1, x._1._2, x._2.open, x._2.high, x._2.low, x._2.close, x._2.openTime, x._2.closeTime) )
        .saveToCassandra("stock", "ohlc_1_min2",
          SomeColumns("symbol", "batch_time", "open", "high", "low", "close", "open_time", "close_time"))

    ssc.start()
    ssc.awaitTermination()
  }

  def reduceOhlc(x: OhlcValue, y: OhlcValue): OhlcValue = {
    new OhlcValue(
      high=math.max(x.high, y.high),
      low=math.min(x.low, y.low),
      open=if(x.openTime.before(y.openTime)) x.open else y.open,
      close=if(x.closeTime.after(y.closeTime)) x.close else y.close,
      openTime=if(x.openTime.before(y.openTime)) x.openTime else y.openTime,
      closeTime=if(x.closeTime.after(y.closeTime)) x.closeTime else y.closeTime
    )
  }

}

case class BatchTransaction(symbol: String, batchTime: Calendar, price: Float, txTime: Calendar)
case class OhlcValue(high: Float, low: Float, open: Float, close: Float, openTime: Calendar, closeTime: Calendar)